coa/pkg/apis/v1alpha2/providers/keylock/memory/memory.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		sLog.Errorf("  P (String Lock): failed to parse provider config %+v", err)
+		return errors.New("expected MemoryKeyLockProviderConfig")
+	}
+	if KeyLockConfig.CleanInterval > 0 {
+		mslp.cleanInterval = KeyLockConfig.CleanInterval
+	} else {
+		mslp.cleanInterval = 30 // default: 30 seconds
+	}
+	if KeyLockConfig.PurgeDuration > 0 {
+		mslp.purgeDuration = KeyLockConfig.PurgeDuration
+	} else {
+		mslp.purgeDuration = 60 * 60 * 12 // default: 12 hours
+	}
+	mslp.lm = NewLockMap()
+	go func() {
+		for {
+			mslp.lm.clean(-mslp.purgeDuration)
+			time.Sleep(time.Duration(mslp.cleanInterval) * time.Second)
+		}
+	}()
+	return nil
+}
+
+func (mslp *MemoryKeyLockProvider) Lock(key string) {
+	mslp.lm.getLockNode(key).Lock()
+}
+
+func (mslp *MemoryKeyLockProvider) UnLock(key string) {
+	mslp.lm.getLockNode(key).Unlock()
+	go mslp.lm.updateLockLRU(key)
+}
+
+func (mslp *MemoryKeyLockProvider) TryLock(key string) bool {
+	return mslp.lm.getLockNode(key).TryLock()
+}
+
+func (mslp *MemoryKeyLockProvider) TryLockWithTimeout(key string, duration time.Duration) bool {
+	start := time.Now()
+	for start.Add(duration).After(time.Now()) {
+		if mslp.TryLock(key) {
+			return true
+		} else {
+			time.Sleep(time.Duration(100) * time.Millisecond)
+		}
+	}
+	return false
+}
+
+type LockNode struct {
+	lastUsedTime time.Time
+	key          string
+	lock         sync.Mutex
+	prev         *LockNode
+	next         *LockNode
+}
+
+type LockMap struct {
+	m    *sync.Map
+	head *LockNode
+	tail *LockNode
+	mu   sync.Mutex
+}
+
+func NewLockMap() *LockMap {
+	lm := LockMap{
+		m:    &sync.Map{},
+		head: nil,
+		tail: nil,
+	}
+	lm.head = &LockNode{prev: nil, next: nil} // dummyhead
+	lm.tail = &LockNode{prev: nil, next: nil} // dummytail
+	lm.head.next = lm.tail
+	lm.tail.prev = lm.head
+	return &lm
+}
+
+func (lm *LockMap) moveToHead(node *LockNode) {
+	if lm.head.next == node {
+		return
+	}
+	if node.prev != nil {
+		node.prev.next = node.next
+	}
+	if node.next != nil {
+		node.next.prev = node.prev
+	}
+
+	node.prev = lm.head
+	node.next = lm.head.next
+	node.next.prev = node
+	lm.head.next = node
+}
+
+func (lm *LockMap) getLockNode(key string) *sync.Mutex {
+	locknode, _ := lm.m.LoadOrStore(key, &LockNode{key: key, prev: nil, next: nil})
+	if ln, ok := locknode.(*LockNode); ok {
+		return &ln.lock
+	} else {
+		print("unexpected behavior")
+		panic("unexpected behavior")
+	}
+}
+
+func (lm *LockMap) updateLockLRU(key string) {
+	lm.mu.Lock()
+	defer lm.mu.Unlock()
+
+	locknode, ok := lm.m.Load(key)
+
+	if ok {
+		locknode.(*LockNode).lastUsedTime = time.Now()
+
+		lm.moveToHead(locknode.(*LockNode))
+	}
+}
+
+func (lm *LockMap) cleanLast(purgeDuration int) bool {
+	if lm.tail.prev == lm.head || lm.tail.prev.lastUsedTime.After(time.Now().Add(time.Duration(purgeDuration)*time.Second)) {
+		return false
+	}
+
+	node := lm.tail.prev
+	if node.prev != nil {
+		node.prev.next = node.next
+	}
+	if node.next != nil {
+		node.next.prev = node.prev
+	}
```

Comment:
node.next should always be lm.tail?

Reply:
yes, the condition can be deleted here after I used a dummy head and tail. Will delete those "if", in the next fix pr

Reply:
sounds good